### PR TITLE
PLANET-2818: social media block

### PR DESCRIPTION
**Pull request by tomisaacmarshall**
Adds basic social media block. This was separated from the video/audio block after discussion with Laura, although I couldn't see a separate jira ticket so have linked to the same one as the video block.

Supports these social media types:
- Twitter: tweet, profile, list, collection, likes, moment
- Facebook: post, activity, photo, video, media, question, note, pages
- Instagram: image

Mostly uses WordPress's built-in oEmbed functionality, except for Facebook pages, as Facebook doesn't offer an oEmbed endpoint for pages.

Selected unsupported social media types (which don't seem to be possible to embed):
- Twitter: hashtag feeds
- Instagram: stories, profiles

A few questions/points:

- Are all blocks meant to have titles? I've added one here
- I have also applied a bottom margin as it seemed this was common to the blocks, but no other styling
- Do we want to add some configuration for alignment (currently the embeds are left aligned) and/or sizing (some of the embeds offer a bit of sizing control)
- We could have further configuration for Facebook pages - currently it just shows the timeline, but you can also choose to show a page's events or messages (or have all three using tabs)
- Some of the iframes take a while to load (e.g. Facebook videos), do we want to add any kind of loading indicator?
- I've left out the `@since` tag I wasn't sure what version it would be

## UPDATE 14/12/18

- Added description field
- Added the three Facebook page options
- Added alignment setting, which uses the standard WP alignment classes used for inline images. So the options are none, left, right, center. (Text will wrap if left or right selected. None is left-aligned with no wrapping.)
- Started to think of these embeds as more akin to an inline image, so changed the margin-bottom to just 1rem, same as you'd see on an inline image. Figured you could use line breaks for more space.
- Had a go at adding a loading spinner but it didn't look that great to me so dropped this for now.
- Haven't added any sizing options for now either

### CODE REVIEW

File: classes/controller/blocks/class-socialmedia-controller.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		/**
+		 * Get all the data that will be needed to render the block correctly.
+		 *
+		 * @param array  $fields This is the array of fields of this block.
+		 * @param string $content This is the post content.
+		 * @param string $shortcode_tag The shortcode tag of this block.
+		 *
+		 * @return array The data to be passed in the View.
+		 */
+		public function prepare_data( $fields, $content = '', $shortcode_tag = 'shortcake_' . self::BLOCK_NAME ): array {
```
Comment (angtheod) [inline classes/controller/blocks/class-socialmedia-controller.php]: Need a space here after closing parenthesis and before :

**Comment by angtheod** [comment]
Code looks nice. Only a minor wp code standard thingy. Also, needs to solve conflict for the styling files. Probably needs to recreate those.

**Comment by angtheod** [comment]
@tomisaacmarshall 
For some of your questions I think we need response from the Issue reporter.
for example:

- Are all blocks meant to have titles?
- Do we want to add some configuration for alignment (currently the embeds are left aligned) and/or sizing (some of the embeds offer a bit of sizing control)
- We could have further configuration for Facebook pages - currently it just shows the timeline, but you can also choose to show a page's events or messages (or have all three using tabs)
- Some of the iframes take a while to load (e.g. Facebook videos), do we want to add any kind of loading indicator?

Do you have access to the relevant Jira [issue](https://jira.greenpeace.org/browse/PLANET-2818)?
I could write those in a comment for you if you don't have rights.

**Comment by tomisaacmarshall** [comment]
> Code looks nice. Only a minor wp code standard thingy. Also, needs to solve conflict for the styling files. Probably needs to recreate those.

Have resolved these

**Comment by tomisaacmarshall** [comment]
> @tomisaacmarshall
> For some of your questions I think we need response from the Issue reporter.
> for example:
> 
> * Are all blocks meant to have titles?
> * Do we want to add some configuration for alignment (currently the embeds are left aligned) and/or sizing (some of the embeds offer a bit of sizing control)
> * We could have further configuration for Facebook pages - currently it just shows the timeline, but you can also choose to show a page's events or messages (or have all three using tabs)
> * Some of the iframes take a while to load (e.g. Facebook videos), do we want to add any kind of loading indicator?
> 
> Do you have access to the relevant Jira [issue](https://jira.greenpeace.org/browse/PLANET-2818)?
> I could write those in a comment for you if you don't have rights.

I don't think I can comment (don't think I have an account), so if you could add that'd be great! I've asked Laura to take a look at the PR comments on rocketchat, though. 

(BTW, maybe I should have added that it would be very trivial to add the alignment and facebook pages options, but possibly a little less straightforward to add size configuration and loading indicator)

**Comment by LauraHilliger** [comment]
> Are all blocks meant to have titles? I've added one here
Yes, all blocks should have a Title and a description field, both optional.

> I have also applied a bottom margin as it seemed this was common to the blocks, but no other styling
@comzeradd can you confirm this is correct on common styling?

> Do we want to add some configuration for alignment (currently the embeds are left aligned) and/or sizing (some of the embeds offer a bit of sizing control)
I tend to say yes, the more options for styling, the better as we have a global and diverse group of web editors. But I have a hard time imagining what exactly you mean. Is there something other than left, right or center aligned? Can you do a text wrap on a block? We haven't done that yet. Same with sizing...we like to limit people's ability to make ugly, but in this case I can't imagine the output 👀 

> We could have further configuration for Facebook pages - currently it just shows the timeline, but you can also choose to show a page's events or messages (or have all three using tabs)
Oh yes, being able to show messages or events will make editors very happy, I think. No tabs.

> Some of the iframes take a while to load (e.g. Facebook videos), do we want to add any kind of loading indicator?
Yes, what kind of advice do you need? We use font awesome for icons if that helps.

File: assets/scss/blocks/_social_media.scss [hunk 2]
```diff
@@ -0,0 +1,16 @@
+.social-media-block {
+  z-index: 3;
+
+  .social-media-embed {
+    margin-bottom: 1rem;
```
Comment (comzeradd) [inline assets/scss/blocks/_social_media.scss]: Can you switch that to `px`? We tend to use `px` for spacing.

**Comment by tamlyn** [comment]
Squashed, rebased and switched to `px`. (Tom's moved on to another project today.)

> The question "What Facebook page content would you like to display?" shouldn't be enabled only when an editor selects "Facebook page" on the "Embed type"?

Sorry, I'm not quite following. There's currently some JS in the admin that hides the field until "Facebook page" is selected. Are you saying that the field should also be disabled in the PHP code, or that the field should not be hidden?

**Comment by comzeradd** [comment]
> Sorry, I'm not quite following. There's currently some JS in the admin that hides the field until "Facebook page" is selected. 

@tamlyn yes, you are right. For some reason it didn't work for me on first run, but it seems to work fine now.

In general it looks good to me. I'll leave it to @angtheod to do a final check and merge.

File: admin/js/blocks-ui.js [hunk 3]
```diff
@@ -169,25 +169,29 @@ jQuery(function ($) {
 
     wp.shortcake.hooks.addAction('shortcode-ui.render_edit', function (shortcodeModel) {
       $(".shortcode-ui-attribute-heading2").parent().before('<p></p>');
+      p4_blocks.social_media.facebook_page_options.init();
```
Comment (kirdia) [inline admin/js/blocks-ui.js]: @tomisaacmarshall please move these to attach_hooks, to keep these hooks in a specific place.
https://github.com/greenpeace/planet4-plugin-blocks/blob/db5e115c26c238c2c30c4c68c7d828e52d7a650a/admin/js/blocks-ui.js#L512

File: admin/js/blocks-ui.js [hunk 4]
```diff
@@ -169,25 +169,29 @@ jQuery(function ($) {
...
     });
 
     wp.shortcake.hooks.addAction('shortcode-ui.render_new', function (shortcodeModel) {
       $(".shortcode-ui-attribute-heading2").parent().before('<p></p>');
+      p4_blocks.social_media.facebook_page_options.init();
```
Comment (kirdia) [inline admin/js/blocks-ui.js]: @tomisaacmarshall same here, please move this to attach_hooks. Or better you can move both lines to attach_hooks and remove this block completely.

https://github.com/greenpeace/planet4-plugin-blocks/blob/db5e115c26c238c2c30c4c68c7d828e52d7a650a/admin/js/blocks-ui.js#L512

**Comment by angtheod** [comment]
Looks good, leaving it to Kyriakos to merge after the final changes he requested.

File: admin/js/blocks-ui.js (hunk 4 above)
Comment (tomisaacmarshall) [reply admin/js/blocks-ui.js]: Thanks, I've now moved these, and tried to make things more consistent with the other blocks, using the attribute hooks in the same manner. However, I wasn't sure how best to initialise the fields, so have maybe gone a bit rogue there!

Also wan't sure what you meant with

> and remove this block completely.

Are you talking about those lines that seem to be adding empty `<p>` tags?

Comment (kirdia) [reply admin/js/blocks-ui.js]: Don't worry about this. it was not introduced in this PR after all.

**Comment by kirdia** [comment]
Looks good to merge! :+1:  @tomisaacmarshall please squash the commits and fix the conflicts to merge it.

**Comment by tomisaacmarshall** [comment]
> Looks good to merge! 👍 @tomisaacmarshall please squash the commits and fix the conflicts to merge it.

Done!

**Pull request merged**